frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/BaseFrontendHandler.java

```diff
@@ -65,44 +70,59 @@ protected BaseFrontendHandler(ConfigurationService configurationService) {
-                                clientSession.ifPresent(
-                                        cs ->
-                                                userContextBuilder.withClient(
-                                                        clientService.getClient(
-                                                                cs
-                                                                        .getAuthRequestParams()
-                                                                        .get("client_id")
-                                                                        .stream()
-                                                                        .findFirst()
-                                                                        .orElseThrow())));
-                                session.ifPresent(
-                                        s ->
-                                                userContextBuilder.withUserProfile(
-                                                        authenticationService
-                                                                .getUserProfileFromEmail(
-                                                                        s.getEmailAddress())));
-
-                                return handleRequestWithUserContext(
-                                        input, context, userContextBuilder.build());
-                            } else {
-                                LOG.error("Session cannot be found");
-                                return generateApiGatewayProxyErrorResponse(
-                                        400, ErrorResponse.ERROR_1000);
-                            }
-                        });
+        return isWarming(input).orElseGet(() -> validateAndHandleRequest(input, context));
     }
 
     public abstract APIGatewayProxyResponseEvent handleRequestWithUserContext(
             APIGatewayProxyRequestEvent input, Context context, UserContext userContext);
+
+    private APIGatewayProxyResponseEvent validateAndHandleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        Optional<Session> session = sessionService.getSessionFromRequestHeaders(input.getHeaders());
+        Optional<ClientSession> clientSession =
+                clientSessionService.getClientSessionFromRequestHeaders(input.getHeaders());
+        if (session.isPresent()) {
+            UserContext.Builder userContextBuilder = UserContext.builder(session.get());
+            clientSession.ifPresent(
+                    cs ->
+                            userContextBuilder.withClient(
+                                    clientService.getClient(
+                                            cs.getAuthRequestParams().get("client_id").stream()
+                                                    .findFirst()
+                                                    .orElseThrow())));
+            session.ifPresent(
```

Comment:
We can probably tidy up some of these optionals like so:

```
session.map(Session::getEmailAddress)
  .map(authenticationService::getUserProfileFromEmail)
  .ifPresent(userContextBuilder::withUserProfile);
```

Reply:
Actually, that particular one on line 93, I was supposed to have deleted as it was superseded by the `ifPresentOrElse()` below.